cmd/kubeadm/app/cmd/phases/kubelet.go

```diff
@@ -111,72 +94,73 @@ var (
-		},
+// NewKubeletStartPhase creates a kubeadm workflow phase that start kubelet on a node.
+func NewKubeletStartPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:    "kubelet-start",
+		Short:   "Writes kubelet settings and (re)starts kubelet",
+		Long:    "Writes a file with KubeletConfiguration and an environment file with node specific kubelet settings, and then (re)starts kubelet.",
+		Example: kubeletStartPhaseExample,
+		Run:     runKubeletStart,
 	}
-
-	options.AddConfigFlag(cmd.Flags(), &cfgPath)
-	return cmd
 }
 
-// RunKubeletWriteEnvFile is the function that is run when "kubeadm phase kubelet write-env-file" is executed
-func RunKubeletWriteEnvFile(cfgPath string) error {
-	internalcfg, err := configutil.AnyConfigFileAndDefaultsToInternal(cfgPath)
-	if err != nil {
-		return err
+// runKubeletStart executes start kubelet logic.
+func runKubeletStart(c workflow.RunData) error {
+	data, ok := c.(kubeletStartData)
+	if !ok {
+		return fmt.Errorf("kubelet-start phase invoked with an invalid data struct")
 	}
 
-	var nodeRegistrationObj *kubeadmapi.NodeRegistrationOptions
-	var featureGates map[string]bool
-	var registerWithTaints bool
-
-	switch cfg := internalcfg.(type) {
-	case *kubeadmapi.InitConfiguration:
-		nodeRegistrationObj = &cfg.NodeRegistration
-		featureGates = cfg.FeatureGates
-		registerWithTaints = false
-	case *kubeadmapi.JoinConfiguration:
-		nodeRegistrationObj = &cfg.NodeRegistration
-		featureGates = cfg.FeatureGates
-		registerWithTaints = true
-	default:
-		return fmt.Errorf("couldn't read config file, no matching kind found")
+	// First off, configure the kubelet. In this short timeframe, kubeadm is trying to stop/restart the kubelet
+	// Try to stop the kubelet service so no race conditions occur when configuring it
+	if !data.DryRun() {
+		glog.V(1).Infof("Stopping the kubelet")
+		kubeletphase.TryStopKubelet()
 	}
 
-	if err := kubeletphase.WriteKubeletDynamicEnvFile(nodeRegistrationObj, featureGates, registerWithTaints, constants.KubeletRunDirectory); err != nil {
+	// Write env file with flags for the kubelet to use. We do not need to write the --register-with-taints for the master,
+	// as we handle that ourselves in the markmaster phase
+	// TODO: Maybe we want to do that some time in the future, in order to remove some logic from the markmaster phase?
+	if err := kubeletphase.WriteKubeletDynamicEnvFile(&data.Cfg().NodeRegistration, data.Cfg().FeatureGates, false, data.KubeletDir()); err != nil {
 		return fmt.Errorf("error writing a dynamic environment file for the kubelet: %v", err)
 	}
+
+	// Write the kubelet configuration file to disk.
+	if err := kubeletphase.WriteConfigToDisk(data.Cfg().ComponentConfigs.Kubelet, data.KubeletDir()); err != nil {
+		return fmt.Errorf("error writing kubelet configuration to disk: %v", err)
+	}
+
+	// Try to start the kubelet service in case it's inactive
+	if !data.DryRun() {
+		glog.V(1).Infof("Starting the kubelet")
+		kubeletphase.TryStartKubelet()
```

Comment:
it's a little weird that this function only prints errors and doesn't return any errors.

kubeadm doesn't really have a way to fail out of this phase. If `TryStartKubelet` fails to start the Kubelet, a message is printed but nothing is returned.

This is probably ok and an existing limitation, but we could bail out of executing the phases earlier if this function returned an error.

Reply:
I agree, I'm opening an issue for changing this here and in the join workflow